File: src/memory/adapters/docdeploy.ts
/**
 * DocDeploy memory connector.
 *
 * DocDeploy is the pay-per-call memory store the Nakatomi project
 * recommends as the default agent memory. Recall via the x402 REST API.
 * Webhook receipts are HMAC-signed with the DOCDEPLOY_WEBHOOK_SECRET
 * (separate from the API key) so inbound auto-linking is verifiable.
 */

import type { Env } from '../../env'
import { hmacSha256Hex } from '../../lib/hmac'
import type { MemoryConnector, MemoryRecallArgs, MemoryRecallResult } from '../types'

const DEFAULT_BASE_URL = 'https://x402.docdeploy.io'

interface DocDeployRecallResponse {
  results?: Array<{
    id?: string
    text?: string
    snippet?: string
    score?: number
    metadata?: Record<string, unknown>
  }>
}

export function docdeployAdapter(env: Env): MemoryConnector {
  const apiKey = (env as unknown as { DOCDEPLOY_API_KEY?: string }).DOCDEPLOY_API_KEY
  const webhookSecret = (env as unknown as { DOCDEPLOY_WEBHOOK_SECRET?: string }).DOCDEPLOY_WEBHOOK_SECRET
  const baseUrl =
    (env as unknown as { DOCDEPLOY_BASE_URL?: string }).DOCDEPLOY_BASE_URL ?? DEFAULT_BASE_URL

  return {
    name: 'docdeploy',
    isConfigured: () => Boolean(apiKey),
    async recall(args: MemoryRecallArgs): Promise<MemoryRecallResult[]> {
      if (!apiKey) return []
      const res = await fetch(`${baseUrl}/v1/recall_memory`, {
        method: 'POST',
        headers: {
          authorization: `Bearer ${apiKey}`,
          'content-type': 'application/json',
        },
        body: JSON.stringify({
          query: args.query,
          limit: args.limit,
          metadata: {
            workspace_id: args.workspaceId,
            crm_entity_type: args.crmEntityType ?? null,
            crm_entity_id: args.crmEntityId ?? null,
          },
        }),
        signal: AbortSignal.timeout(8_000),
      })
      if (!res.ok) return []
      const body = (await res.json()) as DocDeployRecallResponse
      return (body.results ?? []).map((r) => ({
        connector: 'docdeploy',
        external_id: r.id ?? '',
        text: r.text ?? r.snippet ?? '',
        score: typeof r.score === 'number' ? r.score : 0,
        metadata: r.metadata,
      }))
    },
    async verifyWebhook(headers: Headers, rawBody: string): Promise<boolean> {
      if (!webhookSecret) return false
      const sig = headers.get('x-docdeploy-signature')
      if (!sig) return false
      const expected = await hmacSha256Hex(webhookSecret, rawBody)
      return sig === `sha256=${expected}`
    },
  }
}

File: legacy/app/services/memory/adapters/docdeploy.py
"""DocDeploy connector — pay-per-call encrypted memory on Base (x402).

Docs: https://www.docdeploy.io  •  API base: https://x402.docdeploy.io

This is a best-effort adapter based on public docs. Update the request/response
shapes once you have an account against the current API version.
"""

from __future__ import annotations

import logging
import os

import httpx

from app.services.memory.base import MemoryConnector, MemoryItem, MemoryWriteResult

log = logging.getLogger("nakatomi.memory.docdeploy")


class DocDeployConnector(MemoryConnector):
    name = "docdeploy"

    def __init__(self):
        self.api_key = os.getenv("DOCDEPLOY_API_KEY", "")
        self.base_url = os.getenv("DOCDEPLOY_BASE_URL", "https://x402.docdeploy.io").rstrip("/")
        if not self.api_key:
            raise RuntimeError("DOCDEPLOY_API_KEY not set")
        self._client = httpx.Client(
            base_url=self.base_url,
            timeout=10.0,
            headers={"Authorization": f"Bearer {self.api_key}"},
        )

    def store_event(
        self,
        *,
        workspace_id: str,
        event_type: str,
        crm_entity_type: str,
        crm_entity_id: str,
        text: str,
        metadata: dict,
    ) -> MemoryWriteResult | None:
        # Skip noisy events by default; operators can tune.
        if event_type.endswith(".deleted"):
            return None
        payload = {
            "text": text,
            "metadata": {
                **metadata,
                "source": "nakatomi",
                "workspace_id": workspace_id,
                "event_type": event_type,
                "crm_entity_type": crm_entity_type,
                "crm_entity_id": crm_entity_id,
            },
            "tags": [f"nakatomi:{crm_entity_type}", f"event:{event_type}"],
        }
        try:
            r = self._client.post("/v1/remember_memory", json=payload)
            r.raise_for_status()
        except Exception as e:  # noqa: BLE001
            log.warning("docdeploy store failed: %s", e)
            return None
        data = r.json() if r.headers.get("content-type", "").startswith("application/json") else {}
        ext_id = data.get("id") or data.get("memory_id") or ""
        return MemoryWriteResult(connector=self.name, external_id=ext_id, raw_response=data)

    def recall(
        self,
        *,
        workspace_id: str,
        query: str,
        crm_entity_type: str | None = None,
        crm_entity_id: str | None = None,
        limit: int = 10,
    ) -> list[MemoryItem]:
        payload: dict = {"query": query, "limit": limit}
        if crm_entity_type and crm_entity_id:
            payload["filters"] = {
                "metadata.crm_entity_type": crm_entity_type,
                "metadata.crm_entity_id": crm_entity_id,
            }
        try:
            r = self._client.post("/v1/recall_memory", json=payload)
            r.raise_for_status()
            items = r.json().get("items", [])
        except Exception as e:  # noqa: BLE001
            log.warning("docdeploy recall failed: %s", e)
            return []
        out: list[MemoryItem] = []
        for it in items:
            out.append(
                MemoryItem(
                    connector=self.name,
                    external_id=str(it.get("id") or it.get("memory_id") or ""),
                    text=it.get("text") or it.get("content") or "",
                    score=float(it.get("score") or 0.0),
                    metadata=it.get("metadata") or {},
                )
            )
        return out
